File: exp8.py
class Job:
    def __init__(self, id, deadline, profit):
        self.id = id        # Job ID
        self.deadline = deadline  # Deadline of the job
        self.profit = profit  # Profit of the job

def job_sequencing(jobs):
    """
    Solves the Job Sequencing problem using the Greedy strategy.
    
    Parameters:
        jobs: List of Job objects.
        
    Returns:
        Maximum profit and the sequence of jobs.
    """
    # Sort all jobs in descending order of profit
    jobs.sort(key=lambda x: x.profit, reverse=True)

    n = max(job.deadline for job in jobs)  # Find the maximum deadline
    result = [None] * n  # Array to store the result sequence of jobs
    slots = [False] * n  # Boolean array to keep track of free time slots
    total_profit = 0

    # Iterate through all jobs
    for job in jobs:
        # Try to schedule the job in the latest available slot before its deadline
        for j in range(min(n, job.deadline) - 1, -1, -1):
            if not slots[j]:
                slots[j] = True  # Mark the slot as occupied
                result[j] = job.id  # Assign the job to the slot
                total_profit += job.profit  # Add the profit of this job
                break

    # Filter out None values from the result (unscheduled slots)
    scheduled_jobs = [job_id for job_id in result if job_id is not None]
    return total_profit, scheduled_jobs

# Example usage
if __name__ == "__main__":
    jobs = [
        Job("J1", 2, 100),
        Job("J2", 1, 19),
        Job("J3", 2, 27),
        Job("J4", 1, 25),
        Job("J5", 3, 15)
    ]

    max_profit, job_sequence = job_sequencing(jobs)
    print("Maximum Profit:", max_profit)
    print("Job Sequence:", job_sequence)

File: exp8.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Job {
    char id;
    int deadline;
    int profit;
};

bool compare(Job a, Job b) {
    return a.profit > b.profit; // Sort in descending order of profit
}

void jobSequencing(vector<Job>& jobs) {
    int n = jobs.size(), i, j;

    // Sort jobs by profit
    sort(jobs.begin(), jobs.end(), compare);

    // Find maximum deadline to size the slot array
    int maxDeadline = 0;
    for (i = 0; i < n; i++)
        maxDeadline = max(maxDeadline, jobs[i].deadline);

    vector<int> slot(maxDeadline + 1, -1); // -1 means slot is free
    vector<char> result(maxDeadline + 1, ' '); // Store job IDs
    int totalProfit = 0;

    for (i = 0; i < n; i++) {
        for (j = jobs[i].deadline; j > 0; j--) {
            if (slot[j] == -1) {
                slot[j] = i;
                result[j] = jobs[i].id;
                totalProfit += jobs[i].profit;
                break;
            }
        }
    }

    cout << "Scheduled Jobs: ";
    for (i = 1; i <= maxDeadline; i++) {
        if (result[i] != ' ')
            cout << result[i] << " ";
    }
    cout << "\nTotal Profit: " << totalProfit << endl;
}

int main() {
    vector<Job> jobs = {
        {'A', 2, 100},
        {'B', 1, 19},
        {'C', 2, 27},
        {'D', 1, 25},
        {'E', 3, 15}
    };

    jobSequencing(jobs);

    return 0;
}


